feat(resp): optimize simple string "OK" construction

RiversJin opened the pull request:
… pre-allocated memory

By pre-allocating memory, the speed rate of simple strings can be improved. Additionally, using a predefined "+OK\r\n" string eliminates the need for string concatenation when returning "OK".

It solves #2626

mapleFU commented on src/server/redis_reply.cc:
```diff
@@ -37,9 +37,23 @@ namespace redis {
 void Reply(evbuffer *output, const std::string &data) { evbuffer_add(output, data.c_str(), data.length()); }
 
-std::string SimpleString(const std::string &data) { return "+" + data + CRLF; }
+std::string SimpleString(const std::string &data) { 
```
Can the input also be a `std::string_view`?

RiversJin replied:
Of course! In fact, I would really like to switch to using std::string_view as the parameter type. However, I saw that const string& is widely used in the code, so I assumed it was a stylistic choice and didn't make the change. If using the former is acceptable, I’ll make the adjustment later.

mapleFU replied:
> However, I saw that const string& is widely used in the code, so I assumed it was a stylistic choice and didn't make the change.

Previously `+` for string_view would be a bit trick, but now there is no reason not shift to that. For sso string the performance might be same but for long string `string_view` would have better performance

PragmaTwice replied:
For some cases (e.g. we need to pass a null-terminated string to a C api), `std::string` is required since `std::string_view` will not ensure that the string is terminated by `\0`.
In other cases, it can be replaced by std::string_view if the lifetime can be maintained well (no dangling happen).

PragmaTwice replied:
However, previously I was considering to introduce "rope" (a data structure) to build `concat`ed strings in a relatively cheaper way, but it may require many additional work.

mapleFU replied:
> However, previously I was considering to introduce "rope" (a data structure) to build `concat`ed strings in a relatively cheaper way, but it may require many non-trivial work.

It's ok for some network buffer but would be to complex for `SimpleString`...

RiversJin committed:
feat(simplestring): optimize simple string construction speed through pre-allocated memory

By pre-allocating memory, the speed rate of simple strings can be improved.
Additionally, using a predefined "+OK\r\n" string eliminates the need for string concatenation when returning "OK".

mapleFU commented on src/server/redis_reply.cc:
```diff
@@ -37,9 +37,23 @@ namespace redis {
-std::string Error(const Status &s) { return RESP_PREFIX_ERROR + StatusToRedisErrorMsg(s) + CRLF; }
+std::string Error(const Status &s) {
+  std::string res;
+  res.reserve(s.Msg().size() + 3);  // 1 for '-', 2 for CRLF
```
I wonder that `s.Msg()` might not be really the final string generation when `redisErrorPrefixMapping` in `StatusToRedisErrorMsg` contains the specific error message.

RiversJin replied:
Yes, you're absolutely right; I overlooked the possibility that the text could come directly from redisErrorPrefixMapping. However, performing a hash lookup here to confirm the reserved length might be a bit heavy.

Or, do you think it’s a good idea to directly use std::max(16, s.Msg().size() + 3)? In most cases, the minimum allocation unit by the memory allocator is often two pointer lengths, and for the contents within redisErrorPrefixMapping, 16 bytes should be sufficient.

mapleFU replied:
> However, performing a hash lookup here to confirm the reserved length might be a bit heavy.

As the hash table is small and anyway it would be built 😅, I think construct error string firstly is ok...

> do you think it’s a good idea to directly use std::max(16, s.Msg().size() + 3)

It would be good if it's a critical path, but I don't think the error path worth this, since the current code is clear to understand and Error path might not a bottleneck...

RiversJin replied:
Ok, I've reverted this modification.

mapleFU reviewed (commented):
Rest LGTM

mapleFU commented:
Would you mind fix the lint?

RiversJin commented:
> Would you mind fix the lint?

Sure but ...  
From the logs, this is not just a lint issue, but rather due to setting RESP_OK as the type std::string_view. Some function parameters accept const string& instead of std::string_view. If we change everything to accept std::string_view, I think it might require significant modifications...

Since std::string_view cannot be implicitly converted to std::string, and I want to implement RESP_OK as a constexpr variable, I changed it to use constexpr char[].

mapleFU commented:
>  Since std::string_view cannot be implicitly converted to std::string, and I want to implement RESP_OK as a constexpr variable, I changed it to use constexpr char[].

This not fixes the problem, we can either use `const static inline std::string`, or just cast once on "Reply". `const char[]` to `std::string` also applies copy...

RiversJin commented:
> > Since std::string_view cannot be implicitly converted to std::string, and I want to implement RESP_OK as a constexpr variable, I changed it to use constexpr char[].由于 std::string_view 无法隐式转换为 std::string,而且我想将 RESP_OK 实现为 constexpr 变量,因此我将其改为使用 constexpr char[]。
> 
> This not fixes the problem, we can either use `const static inline std::string`, or just cast once on "Reply". `const char[]` to `std::string` also applies copy...这并不能解决问题,我们可以使用 `const static inline std::string` ，或者只在"Reply"上进行一次转换。 `const char[]` 对 `std::string` 也适用于复制...

Ok u're right, I've done

mapleFU commented on src/server/redis_reply.h:
```diff
@@ -39,7 +40,9 @@ namespace redis {
-std::string SimpleString(const std::string &data);
+std::string SimpleString(std::string_view data);
+
+static inline std::string RESP_OK = RESP_PREFIX_SIMPLE_STRING "OK" CRLF;
```
Should also being const?

RiversJin replied:
sorry hhh
